### DISCUSSION TITLE
lsp_workspace_symbols still flashing with sync=true

### DISCUSSION THREAD

**Discussion by cpcopy** [ibhagwan/fzf-lua#1387] [Q&A] [answered]
Run lsp_workspace_symbols with query and the query is not an existing symbol. I can still see the flashing, no matter with global configuration or input as parameter. Is there anything missing?

```
    opts = {
        winopts = {
            split = "belowright new",
        },
        lsp = {
            sync = true
        }
    },
```
or
```
lua require("fzf-lua").lsp_workspace_symbols({lsp_query='test_xx', sync=true})
```
BTW, can the caller know the return status of lsp_workspace_symbols, such as the status of no result?

**Comment by cpcopy** [comment, marked as answer]
Checked the code, have found the option of `sync` was changed to `async`. I think the wiki needs an update. 
Still haven't found the answer for 
> BTW, can the caller know the return status of lsp_workspace_symbols, such as the status of no result?

**Reply by ibhagwan** [reply]
The wiki is open edit if you wanna fix that.

> BTW, can the caller know the return status of lsp_workspace_symbols, such as the status of no result?

The return value is `nil` when no results are found.

**Reply by cpcopy** [reply]
Thanks for such quick response. Done the wiki update. 
I tried following code. The result is always 'nil' though many symbols are returned. 
```
local result = fzfLua.lsp_workspace_symbols({ lsp_query = 't', async = false })
print(result)
```

**Reply by ibhagwan** [reply]
The issue is fzf-lua is run in a coroutine, therefore cannot have a reliable return value, perhaps you can check after exit `if coroutine.running()`?

**Reply by cpcopy** [reply]
I am not familiar with Lua. According to the description of `coroutine.running()`, it is used to indicate which `coroutine` is running, such as main or not. So, it seems cannot indicate 'lsp_workspace_symbols' is still running.
Is there any callback function can used in `lsp_workspace_symbols`? Is it possible to rely on the status of callback to indicate?

**Reply by ibhagwan** [reply]
You can use the `winopts.on_create` lua callback, it will only be called if the interface is open, you can write whatever lua code you want inside it (e.g. set a global variable with `vim.g.xxx`, etc).

**Reply by cpcopy** [reply]
Thanks. It works. Though there is still a timing race between callback and defer function to check the status of callback, it is big progress for my requirement.

**Reply by ibhagwan** [reply]
I can make it return `false` and you can check if the return value `== false` (different from `nil` in lua), let me know if that would help you.

**Reply by cpcopy** [reply]
Do you mean the return value of `lsp_workspace_symbols`, `false` for no result otherwise `true`? That would be helpful. Would there still be the issue caused by `coroutine`?  I mean the return value is not reliable because the job is done `coroutine`.

**Reply by ibhagwan** [reply]
> Do you mean the return value of lsp_workspace_symbols, false for no result otherwise true? That would be helpful.

Correct, unsure about the `true` yet, perhaps it’s better to return the coroutine object.

**Reply by ibhagwan** [reply]
@cpcopy, update to https://github.com/ibhagwan/fzf-lua/commit/61d7f17638285d2bcc18f856d953288a456d0f94, you can now test any fzf-lua funcion to see if it opened the interface with:
```lua
if not require("fzf-lua").lsp_references() then
 ...
end
```

If the function succeeds (i.e. fzf-lua interface is opened) you'll be returned the coroutine thread object:
```lua
local co = require("fzf-lua").fzf_exec("ls")
print("co:", co, "status:", coroutine.status(co))
-- co: thread: 0x41b4ff20 status: suspended
-- **The reason the co is suspended as it waits for the fzf process to terminate
-- which happens when the user either aborted (esc, ctrl-c) or made a selection
```

**Reply by cpcopy** [reply]
Tried it. It works exactly as expected. Thanks so much.
